File: dataset/CoT - Copy/CLAUDE_claude-sonnet-4-5-20250929/Task75.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#ifdef _WIN32
#include <windows.h>
#else
#include <signal.h>
#include <unistd.h>
#include <errno.h>
#endif

/**
 * Validates if string contains only digits
 * @param str Input string
 * @return 1 if valid, 0 otherwise
 */
int isNumeric(const char* str) {
    if (str == NULL || *str == '\\0') {
        return 0;
    }
    
    while (*str) {
        if (!isdigit(*str)) {
            return 0;
        }
        str++;
    }
    return 1;
}

/**
 * Terminates a process by PID with security validations
 * @param pidStr Process ID as string
 * @return 1 if successful, 0 otherwise
 */
int terminateProcess(const char* pidStr) {
    // Input validation - prevent injection attacks
    if (pidStr == NULL || strlen(pidStr) == 0) {
        fprintf(stderr, "Error: PID cannot be null or empty\\n");
        return 0;
    }
    
    // Validate PID is numeric only (prevent command injection)
    if (!isNumeric(pidStr)) {
        fprintf(stderr, "Error: PID must contain only digits\\n");
        return 0;
    }
    
    // Convert to long and validate
    char* endptr;
    long pid = strtol(pidStr, &endptr, 10);
    
    // Check conversion errors
    if (*endptr != '\\0') {
        fprintf(stderr, "Error: Invalid PID format\\n");
        return 0;
    }
    
    // Validate PID is positive
    if (pid <= 0) {
        fprintf(stderr, "Error: PID must be positive\\n");
        return 0;
    }
    
    // Security: prevent terminating system critical PIDs (1-1000)
    if (pid <= 1000) {
        fprintf(stderr, "Warning: Cannot terminate system process (PID <= 1000)\\n");
        return 0;
    }
    
#ifdef _WIN32
    // Windows implementation
    HANDLE hProcess = OpenProcess(PROCESS_TERMINATE, FALSE, (DWORD)pid);
    if (hProcess == NULL) {
        fprintf(stderr, "Error: Cannot open process %ld\\n", pid);
        return 0;
    }
    
    BOOL result = TerminateProcess(hProcess, 1);
    CloseHandle(hProcess);
    
    if (result) {
        printf("Process %ld terminated successfully\\n", pid);
        return 1;
    } else {
        fprintf(stderr, "Failed to terminate process %ld\\n", pid);
        return 0;
    }
#else
    // Unix/Linux implementation
    // Check if process exists
    if (kill(pid, 0) == -1) {
        fprintf(stderr, "Error: Process %ld does not exist or no permission\\n", pid);
        return 0;
    }
    
    // Attempt termination
    if (kill(pid, SIGTERM) == 0) {
        printf("Process %ld terminated successfully\\n", pid);
        return 1;
    } else {
        fprintf(stderr, "Failed to terminate process %ld\\n", pid);
        return 0;
    }
#endif
}

int main(int argc, char* argv[]) {
    printf("=== Process Termination Test Cases ===\\n\\n");
    
    // Test Case 1: Valid PID format
    printf("Test 1: Valid PID (12345)\\n");
    terminateProcess("12345");
    printf("\\n");
    
    // Test Case 2: Invalid PID (negative)
    printf("Test 2: Negative PID (-100)\\n");
    terminateProcess("-100");
    printf("\\n");
    
    // Test Case 3: Invalid PID (non-numeric)
    printf("Test 3: Non-numeric PID (abc123)\\n");
    terminateProcess("abc123");
    printf("\\n");
    
    // Test Case 4: System process protection
    printf("Test 4: System process (PID 1)\\n");
    terminateProcess("1");
    printf("\\n");
    
    // Test Case 5: Empty/null input
    printf("Test 5: Empty PID\\n");
    terminateProcess("");
    printf("\\n");
    
    // Command line argument handling
    if (argc > 1) {
        printf("=== Command Line Execution ===\\n");
        for (int i = 1; i < argc; i++) {
            printf("Attempting to terminate PID: %s\\n", argv[i]);
            terminateProcess(argv[i]);
        }
    }
    
    return 0;
}

File: dataset/CoT/CLAUDE_claude-sonnet-4-5-20250929/Task75.java

import java.io.IOException;

public class Task75 {
    /**
     * Terminates a process by PID with security validations
     * @param pid Process ID to terminate
     * @return true if successful, false otherwise
     */
    public static boolean terminateProcess(String pid) {
        // Input validation - prevent injection attacks
        if (pid == null || pid.trim().isEmpty()) {
            System.err.println("Error: PID cannot be null or empty");
            return false;
        }
        
        // Validate PID is numeric only (prevent command injection)
        if (!pid.matches("^[0-9]+$")) {
            System.err.println("Error: PID must contain only digits");
            return false;
        }
        
        // Convert to long to validate range
        try {
            long pidValue = Long.parseLong(pid);
            if (pidValue <= 0) {
                System.err.println("Error: PID must be positive");
                return false;
            }
            
            // Additional check: prevent terminating system critical PIDs (1-1000)
            if (pidValue <= 1000) {
                System.err.println("Warning: Cannot terminate system process (PID <= 1000)");
                return false;
            }
        } catch (NumberFormatException e) {
            System.err.println("Error: Invalid PID format");
            return false;
        }
        
        try {
            String os = System.getProperty("os.name").toLowerCase();
            ProcessBuilder pb;
            
            if (os.contains("win")) {
                // Windows: Use taskkill with /F flag
                pb = new ProcessBuilder("taskkill", "/F", "/PID", pid);
            } else {
                // Unix/Linux/Mac: Use kill command
                pb = new ProcessBuilder("kill", "-9", pid);
            }
            
            // Security: Redirect error stream to prevent information leakage
            pb.redirectErrorStream(true);
            
            Process process = pb.start();
            int exitCode = process.waitFor();
            
            if (exitCode == 0) {
                System.out.println("Process " + pid + " terminated successfully");
                return true;
            } else {
                System.err.println("Failed to terminate process " + pid);
                return false;
            }
            
        } catch (IOException e) {
            System.err.println("Error: Unable to execute termination command");
            return false;
        } catch (InterruptedException e) {
            System.err.println("Error: Process termination interrupted");
            Thread.currentThread().interrupt();
            return false;
        } catch (SecurityException e) {
            System.err.println("Error: Insufficient permissions");
            return false;
        }
    }
    
    public static void main(String[] args) {
        // Test cases with validation
        System.out.println("=== Process Termination Test Cases ===\\n");
        
        // Test Case 1: Valid PID format
        System.out.println("Test 1: Valid PID (12345)");
        terminateProcess("12345");
        System.out.println();
        
        // Test Case 2: Invalid PID (negative)
        System.out.println("Test 2: Negative PID (-100)");
        terminateProcess("-100");
        System.out.println();
        
        // Test Case 3: Invalid PID (non-numeric)
        System.out.println("Test 3: Non-numeric PID (abc123)");
        terminateProcess("abc123");
        System.out.println();
        
        // Test Case 4: System process protection
        System.out.println("Test 4: System process (PID 1)");
        terminateProcess("1");
        System.out.println();
        
        // Test Case 5: Empty/null input
        System.out.println("Test 5: Empty PID");
        terminateProcess("");
        System.out.println();
        
        // Command line argument handling
        if (args.length > 0) {
            System.out.println("=== Command Line Execution ===");
            for (String pid : args) {
                System.out.println("Attempting to terminate PID: " + pid);
                terminateProcess(pid);
            }
        }
    }
}
